misc-scripts/xref_mapping/XrefParser/Mim2GeneParser.pm

```diff
@@ -19,120 +19,281 @@ limitations under the License.
+                );
+
+ RECORD:
+  while ( my $line = $csv->getline( $eg_io ) ) {
+
+    my ( $is_comment, $is_header )
+      = ( $line->[0] =~ m{
+                           \A
+                           ([#])?
+                           \s*
+                           (MIM[ ]Number)?  # FIXME: this is an assumption regarding header contents.
+                                            # See if $line has split to the right number of columns instead?
+                       }msx );
+    if ( $is_comment ) {
+      if ( ( scalar @{ $line } == $EXPECTED_NUMBER_OF_COLUMNS )
+           && ( ! is_header_file_valid( $line ) ) ) {
+        croak "Malformed or unexpected header in Mim2Gene file '${filename}'";
+      }
+      next RECORD;
     }
-    
-    if((!defined $mim_gene{$omim_id} ) and (!defined $mim_morbid{$omim_id} ) ){
-      $missed_omim++;
-      next;
+
+    if ( scalar @{ $line } != $EXPECTED_NUMBER_OF_COLUMNS ) {
+      croak ' Line ' . $csv->record_number()
+        . " of input file '${filename}' has an incorrect number of columns";
     }
 
-    if($type eq "gene" || $type eq 'gene/phenotype'){
-      if(defined($mim_gene{$omim_id})){
-	foreach my $ent_id (@{$entrez{$entrez_id}}){
-	  foreach my $mim_id (@{$mim_gene{$omim_id}}){
-	    $add_dependent_xref_sth->execute($ent_id, $mim_id);
-	  }
-	}
-	# $add_dependent_xref_sth->execute($entrez{$entrez_id}, $mim_gene{$omim_id});
-      }
-      else{
-	$diff_type++;
-	foreach my $ent_id (@{$entrez{$entrez_id}}){
-	  foreach my $mim_id (@{$mim_morbid{$omim_id}}){
-	    $add_dependent_xref_sth->execute($ent_id, $mim_id);
-	  }
-	}
-	# $add_dependent_xref_sth->execute($entrez{$entrez_id}, $mim_morbid{$omim_id});	
-      }
+    # Do not modify the contents of @{$line}, only the output - hence the /r.
+    my ( $omim_acc, $type, $entrez_id, $hgnc_symbol, $ensembl_id )
+      = map { s{\s+\z}{}rmsx } @{ $line };
+
+    $counters{'all_entries'}++;
+
+    # No point in doing anything if we have no matching MIM xref...
+    if ( ( !defined $mim_gene{$omim_acc} ) &&
+         ( !defined $mim_morbid{$omim_acc} ) )
+    {
+      $counters{'missed_omim'}++;
+      next RECORD;
     }
-    elsif($type eq "phenotype"){
-      if(defined($mim_morbid{$omim_id})){
-	foreach my $ent_id (@{$entrez{$entrez_id}}){
-	  foreach my $mim_id (@{$mim_morbid{$omim_id}}){
-	    $add_dependent_xref_sth->execute($ent_id, $mim_id);
-	  }
-	}
-	#	$add_dependent_xref_sth->execute($entrez{$entrez_id}, $mim_morbid{$omim_id});
-      }
-      else{
-	$diff_type++;
-	foreach my $ent_id (@{$entrez{$entrez_id}}){
-	  foreach my $mim_id (@{$mim_gene{$omim_id}}){
-	    $add_dependent_xref_sth->execute($ent_id, $mim_id);
-	  }
-	}
-	#	$add_dependent_xref_sth->execute($entrez{$entrez_id}, $mim_gene{$omim_id});	
-      }
+
+    # ...or no Ensembl ID or EntrezGene xref to match it to
+    # FIXME: this number might be underestimated because it doesn't
+    # check if Ensembl IDs from the input file actually exist
```

Comment:
And indeed, some of them may not, because of the separation between Ensembl releases (and stable ID updates) and OMIM updates.